Revise update stock logic

mikejolley opened the pull request:
Continuing with #23510 

The current update stock routines mix DB updates and CRUD based saves, so when used within things like bulk update (introduced in https://github.com/woocommerce/woocommerce/commit/6b211ab25bab939ea10752cbc006f293ae5326a9) all havoc breaks loose.

This PR makes some changes to make this more reliable.

- Made the `update_product_stock` routine in the datastore return the new stock amount as well as read then update. The SELECT..FOR UPDATE here might need some feedback from @kloon and @rodrigoprimo 
- The datastore new method `read_stock_quantity` forces a read so we can keep product objects in sync with each other.
- Made `wc_update_product_stock` avoid reads if given a product object
- Allow `wc_update_product_stock` to skip saving if the calling function says it's running updates
- Removed cache clearing from `wc_update_product_stock` since it's done during CRUD save().

Some tests you can do:
- Try the instructions in https://github.com/woocommerce/woocommerce/issues/23388
- Try setting stock to 0 for a bunch of in stock products in bulk. They become out of stock.

@peterfabian If this needs more work or testing (please feel free to take over as I go on leave) it might be better to revert https://github.com/woocommerce/woocommerce/commit/6b211ab25bab939ea10752cbc006f293ae5326a9 instead of rolling out some hacky fixes like in #23510 as those could cause more issues.

Hope it helps :)

peterfabian committed:
Save the product before calling wc_update_product_stock to ensure data consistency and set quantity in local copy after saving it to db in wc_update_product_stock.

wc_update_product_stock reads product from the db twice, so if it's out of sync, strange things are happening.
Also, changes dependant on stock status might get reverted if local copy of $product is not updated after wc_update_product_stock.

mikejolley committed:
Revert double saving change

mikejolley committed:
Improved wc_update_product_stock

Allow product object to be passed through, remove cache clearing which is done during save(), and allow the caller to define whether or not an update routine is already in progress.

mikejolley committed:
Select for update, returning the new stock value

mikejolley committed:
Allow stock to be force read without a change

claudiosanches commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -633,13 +633,13 @@ public function bulk_edit_save( $post_id, $product ) {
 			switch ( $change_stock ) {
 				case 2:
-					wc_update_product_stock( $product, $stock_amount, 'increase' );
+					$new_qty = wc_update_product_stock( $product, $stock_amount, 'increase', true );
```
This variables are not in use in this function.

peterfabian commented on includes/data-stores/class-wc-product-data-store-cpt.php:
```diff
@@ -1291,43 +1304,51 @@ public function get_related_products_query( $cats_array, $tags_array, $exclude_i
+			$current_stock = wc_stock_amount(
+				$wpdb->get_var(
+					$wpdb->prepare(
+						"SELECT meta_value FROM {$wpdb->postmeta} WHERE post_id = %d AND meta_key='_stock' FOR UPDATE;",
```
I think this needs to be wrapped in a transaction, otherwise SELECT in thread 1 will auto commit after the statement and another operation in thread 2 can do SELECT and UPDATE before thread 1 resumes execution and does an UPDATE.

peterfabian commented:
Thanks for picking it up. Just to clarify why I went with the approach I did in #23510: I saw all unit tests for `wc_update_product_stock` saving the product before calling the function, so I assumed that was the intended usage. On the other hand, you're completely right that the way this is currently done does not make much sense--saving changes to local object copy, then (going into `wc_update_product_stock`) re-reading product object from the db twice, saving it again and then patching the copy of the object to be in sync with db (outside of `wc_update_product_stock`) is not robust at all.

peterfabian committed:
Removed unused variables.

peterfabian committed:
Tweaked update lookup table to add the possibility to ignore manage_stock value from the database.

E.g. if manage_stock property has been changed in the object/local copy of the product while updating it, but the product has not been saved to the db yet, the lookup table update would ignore stock quantity update.

peterfabian committed:
Made the increase and decrease operations atomic, set is atomic as it's not fetch-and-update, just update.

peterfabian committed:
Added some comments to hopefully make the function a bit clearer.

kloon reviewed (commented):
This seems to resolve the issue at hand for me to some degree. When a product is a variable one and the variations and stock is not managed on a variation level and you use the bulk edit on the parent which has out of stock variations it does not update the stock status from out of stock to in stock.

Perhaps how it should be, if go and edit the variable product variation and set them all to stock status in stock and save it updates fine. My understanding on the bulk edit is that it should only edit the parent right? In that case it is not an issue.

rodrigoprimo commented on includes/data-stores/class-wc-product-data-store-cpt.php:
```diff
@@ -1285,61 +1298,112 @@ public function get_related_products_query( $cats_array, $tags_array, $exclude_i
+		} else {
+			// 'fetch-and-update' atomic operation for stock quantity.
+			try {
+				wc_transaction_query( 'start' );
```
I believe it would be prudent to get feedback from someone with more experience with databases before proceeding with this approach. In the past, we had to revert commits that used transactions due to unexpected issues (#18175 and #20537).

My concern here is that there is some application time between the `SELECT FOR UPDATE` and the `UPDATE` that is executed in `WC_Product_Data_Store_CPT::set_product_stock()` and that could cause the locks on the rows to block other clients who are running the same queries. I imagine this could be a problem for busy stores, as I think this transaction is executed whenever an order that includes a stock managed product is placed.

peterfabian replied:
> My concern here is that there is some application time between the SELECT FOR UPDATE and the UPDATE that is executed in WC_Product_Data_Store_CPT::set_product_stock() and that could cause the locks on the rows to block other clients who are running the same queries.
Well, that's the purpose of a lock, so I would hope so. I was hoping to make the query non blocking by using NOWAIT, but seems like that's not an option. 

In addition, reading through [MySQL 5.5 manual](https://dev.mysql.com/doc/refman/5.5/en/innodb-deadlock-detection.html), I can see that InnoDB has some other issues as well:
> When InnoDB performs a complete rollback of a transaction, all locks set by the transaction are released. However, if just a single SQL statement is rolled back as a result of an error, some of the locks set by the statement may be preserved. This happens because InnoDB stores row locks in a format such that it cannot know afterward which lock was set by which statement.

I've removed the locks in af79509.

rodrigoprimo commented on includes/data-stores/class-wc-data-store-wp.php:
```diff
@@ -555,9 +555,10 @@ protected function get_search_stopwords() {
 	 * @return array
 	 */
-	protected function get_data_for_lookup_table( $id, $table ) {
+	protected function get_data_for_lookup_table( $id, $table, $ignore_manage_stock = false ) {
```
It feels wrong to add `$ignore_manage_stock` here as this parameter only makes sense in the context of products. Should we maybe create a more generic parameter like `array $options` or maybe we can get rid of this parameter and always ignore the value of manage_stock and update the stock quantity?

peterfabian replied:
Perhaps you're right, I was thinking about adding a more general attribute, but then changed it back as lots of times I saw that planning for future that never comes just introduces more unnecessary abstractions. However, I went with your other suggestion, seems like ignoring the db value of manage_stock should be fine. Updated in 68a3252.

rodrigoprimo commented on includes/data-stores/class-wc-data-store-wp.php:
```diff
@@ -578,8 +579,11 @@ protected function get_primary_key_for_lookup_table( $table ) {
+	 * @return NULL
 	 */
-	protected function update_lookup_table( $id, $table ) {
+	protected function update_lookup_table( $id, $table, $force_qty_update = false ) {
```
Same as above

rodrigoprimo reviewed (changes requested):
I could confirm that this PR fixes bulk editing product stock but I found a problem in one of the queries that need to be addressed before we merge this PR. I also left a few other comments with some suggestions for consideration.

peterfabian committed:
Removed the extra parameter that didn't make sense.

Also get_data_for_lookup_table should not be checking db value of  _manage_stock as that can be in the process of being changed.

peterfabian committed:
Removed the transaction and locking code.

NOWAIT is not supported on MySQL server 5.5,

SELECT ... FOR UPDATE can leave lingering locks on InnoDB ("When InnoDB performs a complete rollback of a transaction, all locks set by the transaction are released. However, if just a single SQL statement is rolled back as a result of an error, some of the locks set by the statement may be preserved. This happens because InnoDB stores row locks in a format such that it cannot know afterward which lock was set by which statement.", via https://dev.mysql.com/doc/refman/5.5/en/innodb-deadlock-detection.html)

peterfabian commented on includes/data-stores/class-wc-product-data-store-cpt.php:
```diff
@@ -1285,61 +1298,112 @@ public function get_related_products_query( $cats_array, $tags_array, $exclude_i
+				$current_stock = wc_stock_amount(
+					$wpdb->get_var(
+						$wpdb->prepare(
+							"SELECT meta_value FROM {$wpdb->postmeta} WHERE post_id = %d AND meta_key='_stock' FOR UPDATE NOWAIT;",
```
Well spotted, thanks. Removed locking.

peterfabian committed:
Removed comments that are no longer relevant.

peterfabian committed:
Foxed the use to correct stock quantity parameter.

rodrigoprimo commented on includes/data-stores/class-wc-product-data-store-cpt.php:
```diff
@@ -1939,8 +1996,7 @@ public function query( $query_vars ) {
 			$price_meta   = (array) get_post_meta( $id, '_price', false );
-			$manage_stock = get_post_meta( $id, '_manage_stock', true );
-			$stock        = 'yes' === $manage_stock ? wc_stock_amount( get_post_meta( $id, '_stock', true ) ) : null;
+			$stock        = wc_stock_amount( get_post_meta( $id, '_stock', true ) );
```
I believe we still need the `_manage_stock` check as, if stock is not managed, we want to set `$stock` to `null` instead of `0`? I haven't checked all the details, but this is probably needed to differentiate products that are out of stock and where the stock is not managed.

peterfabian replied:
Haven't seen it used anywhere to take advantage of this difference, but maybe I missed something and it's safer to not change it.
With the introduction of other changes in this PR, it seems like this change is no longer needed, so I put back the manage_stock check.

rodrigoprimo replied:
> Haven't seen it used anywhere to take advantage of this difference

Both the report and the dashboard widget that displays out of stock products depend on `null` being used for products that don't have managed stock enabled to work properly. For example, here is the query used in the dashboard widget:

https://github.com/woocommerce/woocommerce/blob/10abe6d4be9351d9bcc86db1afe5264567786221/includes/admin/class-wc-admin-dashboard.php#L232

peterfabian replied:
You're right, sorry, I can see that would cause problems now. Thanks for spotting it.

rodrigoprimo reviewed (commented):
Thanks for your changes. I believe they fix all the issues that I mentioned in the previous review. I found just one more thing that could be a problem and left an in-line comment.

peterfabian committed:
Put back the manage stock check for lookup table.

kloon reviewed (approved):
With adding back the stock management check I think we are good to go with this.
